### PR TITLE
feat: move Java Owlbot into this repository for postprocessing

### PR DESCRIPTION
**Pull request by diegomarquezp**
This is stage 1 of the Owlbot replacement project.
This PR moves the implementation of Synthtool's Java Owlbot to this repository.

## Changes
 - Synthtool will be used as an external tool by cloning it and running `pip install`
 - The Java Owlbot [implementation](https://github.com/googleapis/synthtool/tree/59fe44fde9866a26e7ee4e4450fd79f67f8cf599/docker/owlbot/java) has been transferred to `sdk-platform-java`
 - The non-java-specific Owlbot CLI (copy-code) remains referenced as a docker image
 - The interface of `postprocess_library.sh` will be modified (see next section). It will still, however, be intended to be called by `generate_library.sh`. It will still be expected to be run on pre-existing fully functional GAPIC libraries.

### New interface of `postprocess_library.sh`

We will define the interface of `postprocess_library.sh` as follows:

**Input parameters**

<table>
  <tr>
   <td><strong>Argument</strong>
   </td>
   <td><strong>Required</strong>
   </td>
   <td><strong>Description</strong>
   </td>
  </tr>
  <tr>
   <td>postprocessing_target
   </td>
   <td>yes
   </td>
   <td>This is the main parameter to the script. Java Owlbot runs its scripts in this target directory and modifies it (hence this being also an output). Some files are expected to be here<ul>

<li><code>owlbot.py</code> used by java-owlbot
<li><code>.OwlBot.yaml</code> (optional) used to transfer code from <code>preprocessed_sources_path</code>
<li><code>.repo-metadata.json</code> - essential metadata file used by java owlbot

<p>
In the context of <code>generate_library.sh</code>, <code>destination_path</code> should be pointing to a folder with such files, since it is <code>destination_path</code> that will be called as <code>postprocessing_target</code></li></ul>

   </td>
  </tr>
  <tr>
   <td>versions_file
   </td>
   <td>yes
   </td>
   <td>Versions to be applied to README.md and pom files. This is done by java owlbot
   </td>
  </tr>
  <tr>
   <td>preprocessed_sources_path
   </td>
   <td>yes
   </td>
   <td>This folder is the input preprocessed libraries, which should look like a googleapis-gen folder (<a href="https://github.com/googleapis/googleapis-gen/tree/master/google/cloud/asset/v1/google-cloud-asset-v1-java">example</a>). The script will infer the necessary `proto_path` variable from the proto library. 
.
<p>
In the perspective of <code>generate_library.sh</code>, the script will generate a temporary folder with such structure and will pass it to <code>postprocess_library.sh</code> 
   </td>
  </tr>
</table>

**Outputs**

<table>
  <tr>
   <td><strong>Artifact</strong>
   </td>
   <td><strong>Description</strong>
   </td>
  </tr>
  <tr>
   <td>postprocessing_target
   </td>
   <td>Also passed as input to the script. The post-processor will modify the files contained here. It will also use copy-code to transfer preprocessed java sources into this folder if <code>preprocessed_sources_path</code> is provided..
   </td>
  </tr>
</table>

## Remarks
 - Since `synthtool` and `java owlbot` run on python, a python version will be stored in `library_generation/configuration`. This will be managed via `pyenv`
 - Since synthtool [is designed to run on Linux _only_](https://github.com/googleapis/synthtool/blob/ea6f80056a7d22f4d3a3e8fee2d59cdc746470bd/synthtool/README.md?plain=1#L12), the integration test with post-processing for macos were disabled. We _will not_ follow up in a later stage by creating a synthtool-only image ([POC](https://github.com/googleapis/synthtool/pull/1902))
 - `google-java-format` was originally [added in the image definition](https://github.com/googleapis/synthtool/blob/ea6f80056a7d22f4d3a3e8fee2d59cdc746470bd/docker/owlbot/java/Dockerfile#L32). The scripts will now [download it at runtime](https://github.com/googleapis/sdk-platform-java/blob/9d4a54ca62fa88a12e8e2cb1fefa1a31300db150/library_generation/owlbot/bin/format_source.sh#L43-L48). The formatter requires java 11, so the github integration test was updated to run on java 11.
 - Since `destination_path` was previously being used as a pre-existing, fully functional GAPIC library when postprocessing is enabled, it will now clash with the java owlbot postprocessor by having unexpected folders in the GAPIC folder (i.e. preprocessed libraries produced by `generate_library.sh` will be sent to, for example, `google-cloud-java/java-.../` which will produce a library with pre-processed folders). This is dealt with by having a `temp_destination_path` in `generate_library.sh` that we will either (1) use it as `preprocessed_sources_path` when postprocessing is enabled or (2) simply transfer its contents to `destination_path` when it's not enabled.
 - Due to https://github.com/googleapis/sdk-platform-java/issues/2303, I temporarily disabled the tests for `google/cloud/asset/v1p2beta1` and `google/cloud/compute`

### CODE REVIEW

File: .github/workflows/verify_library_generation.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
   integration_tests:
     strategy:
       matrix:
-        java: [ 8 ]
+        java: [ 11 ]
```
Comment (diegomarquezp) [inline .github/workflows/verify_library_generation.yaml:15]: This is for google-java-format

File: library_generation/owlbot/bin/format_source.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+# Copyright 2021 Google LLC
```
Comment (JoeWang1127) [inline library_generation/owlbot/bin/format_source.sh:2]: Please change the year.

File: library_generation/test/resources/proto_path_list.txt [hunk 3]
```diff
@@ -5,13 +5,13 @@
 # google/bigtable/v2 java-bigtable true
 google/cloud/apigeeconnect/v1 java-apigee-connect false
 google/cloud/asset/v1p5beta1 java-asset false
-google/cloud/asset/v1p2beta1 java-asset false
+# google/cloud/asset/v1p2beta1 java-asset false
```
Comment (JoeWang1127) [inline library_generation/test/resources/proto_path_list.txt:8]: Why do we disable this test?

Comment (diegomarquezp) [reply library_generation/test/resources/proto_path_list.txt:8]: I forgot to add a mention of https://github.com/googleapis/sdk-platform-java/issues/2303 - I will add it to the description

Some googleapis protos get updated before they appear in googleapis-gen or google-cloud-java

File: library_generation/owlbot/bin/format_source.sh (hunk 2 above)
Comment (diegomarquezp) [reply library_generation/owlbot/bin/format_source.sh]: done

File: .github/workflows/verify_library_generation.yaml [hunk 4]
```diff
@@ -36,17 +51,10 @@ jobs:
...
-    - name: install docker (macos)
-      if: matrix.os == 'macos-12'
-      run: |
-        brew update --preinstall
-        brew install docker docker-compose qemu 
-        brew upgrade qemu
-        colima start
-        docker run --user $(id -u):$(id -g) --rm hello-world
     - name: Run integration tests
+      if: matrix.os == 'ubuntu-22.04' || matrix.post_processing == 'false'
```
Comment (suztomo) [inline .github/workflows/verify_library_generation.yaml:55]: Can you add a reason why you don't run the integration tests in this condition?

**Review by suztomo — APPROVED** [review]
Memo: As per the design doc, we won't use this until the end of Q1 2024.

File: .github/workflows/verify_library_generation.yaml (hunk 4 above)
Comment (diegomarquezp) [reply .github/workflows/verify_library_generation.yaml:56]: done

File: library_generation/owlbot/bin/format_source.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    echo "File skipped formatting: $file"
+  else
+   echo $file >> $tmpfile
+  fi
+done
+
+# download the google-java-format tool
+if [ ! -f "${scripts_root}/owlbot/google-java-format.jar" ]; then
+  echo 'downloading google-java-format'
+  java_format_version=$(cat "${scripts_root}/configuration/java-format-version")
```
[the remark this answers is not in the document]
Comment (suztomo) [reply library_generation/owlbot/bin/format_source.sh:45]: Owlbot postprocessors run in a GCP project (Automation team's Cloud Build) and the automation team did not want to run untrusted code (such as maven plugins defined in pom.xml) in a pull request.

File: library_generation/owlbot/templates/poms/bom_pom.xml.j2 [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    BOM for {{name}}
+  </description>
+
+  <organization>
+    <name>Google LLC</name>
+  </organization>
+
+  <developers>
+    <developer>
+      <id>chingor13</id>
```
[the remark this answers is not in the document]
Comment (suztomo) [reply library_generation/owlbot/templates/poms/bom_pom.xml.j2:26]: Yes, it's always good to update the information.

**Review by suztomo — COMMENTED** [review]
Until we start using this postprocessor (around Feb/March 2024), people, including me, continue make changes to the Owlbot Java postprocessor via the synthtool repository. Would you write some README section what steps are needed to copy these changes in future (when?)?

File: library_generation/owlbot/bin/format_source.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply library_generation/owlbot/bin/format_source.sh:45]: I see, now that we are moving the script to sdk-platform-java, can we just run `mvn fmt:format` @diegomarquezp ?

[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/bin/format_source.sh:45]: Ack

File: library_generation/owlbot/templates/poms/bom_pom.xml.j2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/templates/poms/bom_pom.xml.j2:26]: [The new library scripts](https://github.com/googleapis/google-cloud-java/blob/de8046bfe66a99c7ef3c5205359ae5f11f74a865/generation/consolidate_config.sh#L77C1-L86) remove this and other sections in the final GAPIC library.

Maybe we can just remove this from the templates?

File: library_generation/owlbot/templates/poms/cloud_pom.xml.j2 [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+      <groupId>org.threeten</groupId>
+      <artifactId>threetenbp</artifactId>
+    </dependency>
+
+    <!-- Test dependencies -->
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+      <version>4.13.2</version>
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/templates/poms/cloud_pom.xml.j2:87]: It is removed by the new library scripts - [it never reaches our final libraries](https://github.com/search?q=repo%3Agoogleapis%2Fgoogle-cloud-java+4.13.2&type=code)

I think we can remove this one and other elements of the template (see https://github.com/googleapis/sdk-platform-java/pull/2282/files#r1428199215)

File: library_generation/owlbot/templates/poms/bom_pom.xml.j2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply library_generation/owlbot/templates/poms/bom_pom.xml.j2:26]: > Maybe we can just remove this from the templates?

Yep, SGTM.

[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/templates/poms/bom_pom.xml.j2]: Done

File: library_generation/owlbot/templates/poms/cloud_pom.xml.j2 (hunk 7 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/templates/poms/cloud_pom.xml.j2:87]: Done

File: library_generation/owlbot/bin/format_source.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/owlbot/bin/format_source.sh]: Done

File: library_generation/configuration/java-format-version
Comment (blakeli0) [inline library_generation/configuration/java-format-version:1]: I think we could remove this file now.

**Comment by blakeli0** [comment]
> Until we start using this postprocessor (around Feb/March 2024), people, including me, continue make changes to the Owlbot Java postprocessor via the synthtool repository. Would you write some README section what steps are needed to copy these changes in future (when?)?

That's valid concern. How often do we usually update owlbot post-processor? We thought it does not change often so expected it to not change in another two months. But if it does change, I think we can either 
1. Update the post-processing logic here as well when we make changes to synthtool. 
2. Don't do copy the changes over until the final cut over, we can copy all the changes in the past few months at that time in one shot.

**Review by blakeli0 — APPROVED** [review]
LGTM. As we discussed offline, please look into the possibility of moving the Java specific Github templates into this repo, but it can be a separate PR.

**Review by meltsufin — APPROVED** [review]
Nice work @diegomarquezp!

**Comment by diegomarquezp** [comment]
> Until we start using this postprocessor (around Feb/March 2024), people, including me, continue make changes to the Owlbot Java postprocessor via the synthtool repository. Would you write some README section what steps are needed to copy these changes in future (when?)?

@suztomo Done

File: library_generation/configuration/java-format-version
Comment (diegomarquezp) [reply library_generation/configuration/java-format-version:1]: removed

**Comment by diegomarquezp** [comment]
> > Until we start using this postprocessor (around Feb/March 2024), people, including me, continue make changes to the Owlbot Java postprocessor via the synthtool repository. Would you write some README section what steps are needed to copy these changes in future (when?)?
> 
> That's valid concern. How often do we usually update owlbot post-processor? We thought it does not change often so expected it to not change in another two months. But if it does change, I think we can either
> 
> 1. Update the post-processing logic here as well when we make changes to synthtool.
> 2. Don't do copy the changes over until the final cut over, we can copy all the changes in the past few months at that time in one shot.

@blakeli0 I missed point `2` when modifying the README. I mentioned in the README that there were slight tweaks here and there in the transferred owlbot implementation. Whether we modify them within or after the two coming months, the instructions apply to both cases equally.

**Comment by diegomarquezp** [comment]
IT for `google/cloud/asset/v1` failing as recent changes in googleapis have not been reflected in google-cloud-java (pending PR: https://github.com/googleapis/google-cloud-java/pull/10138)

**Pull request merged**